🏛️ Implement new column picker

Droniu opened the pull request:
I want to merge this change because it adds new column picker in datagrid views. Fixes #3270.
<img width="708" alt="image" src="https://user-images.githubusercontent.com/41952692/232475266-fb22c710-38bb-495e-ba5b-b1fd7e63068f.png">

Macaw-ui prerequirements:
- [x] Toggle https://github.com/saleor/macaw-ui/pull/328
- [x] Checkbox https://github.com/saleor/macaw-ui/pull/274
- [x] Popover https://github.com/saleor/macaw-ui/pull/335

To-do:
- [x] Create new column architecture
- [x] Create column picker component
    - [x] Static column section
    - [x] Custom column section
    - [x] Available column section
    - [x] Pagination
- [x] Implement parsers for product list
- [x] Implement parsers for order list (afer rebasing)
- [x] Remove legacy code
- [x] Refactor naming & clean-up
- [x] Stories
- [x] Add unit tests
- [x] Add markdown docs for architecture & new view adaptation
- [x] Preselect category if there is only one
- [x] Scroll to new column
- [x] Optimistic UI on dynamic columns
- [x] At least 1 static column should be always present


In-development bugs:
- [x] Fix displaying custom column value
- [x] Fix custom column sorting
- [x] Fix jumpy behaviour of custom column section
- [x] Portalize popover, fixing scroll issue for larger lists of custom columns & border radius clipping https://github.com/saleor/macaw-ui/pull/394
- [x] Adding available column to custom section closes the modal when it shouldn't
- [x] Missing cursor pointer on available columns https://github.com/saleor/macaw-ui/pull/417
- [x] Add min-height to avoid jumping inside the popover
- [x] Attribute columns have unexpected padding
- [x] Add margin to right side of modal (match top padding)
- [x] Column picker button should be in pressed state
- [x] Expanded state of picker shouldn't be persisted when it's closed
- [x] Long dynamic column names break line - make this section growable

Will be worked later on:
- #3686
- #3656 
- #3657
- #3682 
- #3685 

### New column architecture
<img width="977" alt="image" src="https://user-images.githubusercontent.com/41952692/233042483-d2cb30f3-26b7-40b5-9d08-2ea42f7f0242.png">

### Acceptance Criteria
1. Column picker has two types of columns: static and dynamic.
2. Static columns are predefined - for example id, name, description.
3. Dynamic columns can be categorized into groups, for example attributes, stocks, channels. They cannot be predefined because they come from the API. Example of dynamic column is Asia (warehouse) or Color (attribute).
4. User should be able to toggle columns inside the picker, which is instantly reflected in the datagrid. Toggling column to ON scrolls horizontally to this column in the datagrid.
5. After clicking plus button in the picker, user should be able to select dynamic column category (or it is preselected when there is only one category). Then, user can search through dynamic column candidates and add them as togglable dynamic column. Candidates should be displayed as paginated list.
6. State of the selected columns is persisted in local storage. Column picker should work when local storage is empty.
7. At least one static column should always be visible.
8. Some of the columns are sortable - we shouldn't introduce any regression here.
9. User should be able to change width of the columns in the datagrid.
10. User should be able to change order of the columns in the datagrid.
11. UI of column picker should be optimistic. For example adding dynamic column candidate should happen instantly.
12. Column picker should be implemented in **product list** where there are 6 static columns: product, type, description, availability, last updated, price and 1 dynamic column category: attributes.
13. Column picker should be implemented in **order list** where there are 6 static columns: order number, date, customer, payment status, fulfillment status, total price. There are no dynamic columns in this view.
14. Amount of GrahpQL queries should be minimized.
15. UX should be seamless. There shouldn't be any unexpected jumps, delays and components have to follow quality requirements. e.g. all clickable components should have cursor: pointer.
16. Column picker component should match designs provided in Figma.


**PR intended to be tested with API branch:** main

### Pull Request Checklist

<!-- Please keep this section. It will make maintainer's life easier. -->

1. [x] This code contains UI changes
17. [x] All visible strings are translated with proper context including data-formatting
18. [x] Attributes `[data-test-id]` are added for new elements
19. [ ] Changes are mentioned in the changelog
20. [x] The changes are tested in different browsers and in light/dark mode

### Test environment config

<!-- Do not remove this section. It is required to properly setup test deployment instance.
Modify API_URI if you want test instance to use custom backend. CYPRESS_API_URI is optional, use when necessary. -->

API_URI=https://automation-dashboard.staging.saleor.cloud/graphql/
APPS_MARKETPLACE_API_URI=https://apps.staging.saleor.io/api/v2/saleor-apps

### Do you want to run more stable tests?

To run all tests, just select the stable checkbox. To speed up tests, increase the number of containers. Tests will be re-run only when the "run e2e" label is added.

1. [x] stable
2. [ ] giftCard
3. [ ] category
4. [ ] collection
5. [ ] attribute
21. [ ] productType
22. [ ] shipping
23. [ ] customer
24. [ ] permissions
25. [ ] menuNavigation
26. [ ] pages
27. [ ] sales
28. [ ] vouchers
29. [ ] homePage
30. [ ] login
31. [ ] orders
32. [ ] products
33. [ ] app
34. [ ] plugins
35. [ ] translations
36. [ ] navigation
37. [ ] variants
38. [ ] payments

CONTAINERS=4

poulch commented on src/components/ColumnPicker/useColumns.ts:
```diff
@@ -0,0 +1,102 @@
+  );
+
+  const [visibleColumns, setVisibleColumns] =
+    useStateFromProps(initialColumnsState);
```
suggestion: bro this is deprecated

Droniu replied:
We deprecated it because it was used excessively in the dashboard, however in this case this is exactly what is needed - useState with state update based on deep comparisons. Reference update (useStateUpdate) will lead to infinite loop there and shallow comparisons won't work too because changes are deep in the object (and hook like this is not present in the dashboard too). We might think about undoing the depreciation on this hook, just with warning about its usage, because it's heavy artillery and usually there are better solutions.

poulch commented on src/components/ColumnPicker/useColumns.ts:
```diff
@@ -0,0 +1,102 @@
+  const [visibleColumns, setVisibleColumns] =
+    useStateFromProps(initialColumnsState);
+
+  const onMove = React.useCallback(
+    (startIndex: number, endIndex: number): void => {
+      setVisibleColumns(old =>
+        addAtIndex(old[startIndex], removeAtIndex(old, startIndex), endIndex),
+      );
+    },
+    [setVisibleColumns],
```
question: should you handle empty column here, like don't allow to move normal column behind empty?

Droniu replied:
Great catch.

poulch replied:
Still don't see empty column handling

Droniu replied:
Fixed now.

poulch commented on src/components/ColumnPicker/utils.tsx:
```diff
@@ -0,0 +1,43 @@
+  if (columnCategories.length === 1) {
+    return <CloseIcon />;
+  }
+  if (selectedCategory) {
+    return <ArrowLeftIcon />;
+  } else {
+    return <CloseIcon />;
+  }
```
```suggestion
  if (selectedCategory) {
    return <ArrowLeftIcon />;
  }
  
   return <CloseIcon />;
```

Droniu commented on src/components/ColumnPicker/NewColumPickerCategories.tsx:
```diff
@@ -0,0 +1,159 @@
+  );
+
+  const changeHandler = (column: string) => {
+    if (columnPickerSettings.includes(column)) {
+      onDynamicColumnSelect(
+        columnPickerSettings.filter(currentCol => currentCol !== column),
+      );
+    } else {
+      onDynamicColumnSelect([...columnPickerSettings, column]);
+    }
```
Might look better but ternary is an expression, which fails eslint check (no unused expressions). I could use return here but I would need to test it on view with multiple categories, and we don't have a view like that for now, so I'm gonna leave it as it is

Droniu commented on src/components/ColumnPicker/NewColumPickerCategories.tsx:
```diff
@@ -0,0 +1,159 @@
+    }
+  };
+
+  React.useEffect(() => {
+    // Preselect category when there is only one
+    if (columnCategories.length === 1) {
+      setSelectedCategory(columnCategories[0].name);
+    }
+  }, [columnCategories]);
```
Column categories come from API so they might be undefined at first, I haven't implemented it with strict nulls in mind. If I set this as initial value of useState it might not work when component is rendered without data from the API

Droniu commented on src/components/ColumnPicker/useColumnPickerSettings.ts:
```diff
@@ -0,0 +1,32 @@
+
+const COLUMN_PICKER_KEY = "columnPickerConfig";
+
+export type DatagridViews = "PRODUCT_LIST" | "PRODUCT_DETAILS" | "ORDER_LIST";
```
I didn't know about order details and draft details while implementing this, I'll add them. This will have to be modified when adapting new views.

Droniu commented on src/components/ColumnPicker/ColumnPicker.md:
````diff
@@ -0,0 +1,162 @@
+  },
+];
+```
+Here we only have 1 column category, attributes. `attributesData` is the result of the first query, `gridAttributesData` is the result of the second query. We also provide pagination props, which are used in the column picker.
````
Reasons why we use two queries are explained in lines 100-104. I will add examples of queries.

poulch commented on src/components/ColumnPicker/useColumns.ts:
```diff
@@ -0,0 +1,102 @@
+}: UseColumnsProps) => {
+  const dynamicColumns: AvailableColumn[] = React.useMemo(
+    () =>
+      columnCategories.reduce((columns: AvailableColumn[], category) => {
+        category.selectedNodes.forEach(column => {
+          if (columnPickerSettings.includes(column.id)) {
+            columns.push(column);
+          }
+        });
+        return columns;
+      }, [] as AvailableColumn[]),
```
nitpick: 
```suggestion
      columnCategories.reduce<AvailableColumn[]>((columns: AvailableColumn[], category) => {
        category.selectedNodes.forEach(column => {
          if (columnPickerSettings.includes(column.id)) {
            columns.push(column);
          }
        });
        return columns;
      }, []),
```

poulch reviewed (dismissed):
Looks good. I've found a couple of places to improve but not mandatory:
1. Can you split components into smaller ones, for example, NewColumnPicker
2. Why columnPickerSettings aren't used directly inside each datagrid component but passed through two components down

Droniu commented:
@poulch Why columnPickerSettings aren't used directly inside each datagrid component but passed through two components down - this is explained in docs in line 50:

> The reason why column picker settings object needs to be in the view file and cannot be integrated into internal logic of useColumns is because we use column picker settings in the query. We need to know which columns are selected in order to fetch the correct data from the API.

Droniu commented on src/orders/components/OrderListDatagrid/datagrid.ts:
```diff
@@ -26,62 +24,48 @@ import {
-
-  return columns;
-};
+export const orderListStaticColumnAdapter = (emptyColumn, intl, sort) => [
```
Yep, it's good idea to memoize them - both for order and product list

Droniu commented on src/products/components/ProductListDatagrid/ProductListDatagrid.tsx:
```diff
@@ -71,104 +81,133 @@ export const ProductListDatagrid: React.FC<ProductListDatagridProps> = ({
-    ],
-  );
+    );
+  }, [activeAttributeSortId, sort]);
```
Can't do this :( Creates infinite loop for some reason - we would probably need more useCallbacks inside useColumns. This useEffect will be dropped in #3685 anyway, so I'm inclined to skip this problem for now.

Droniu commented on src/orders/components/OrderListDatagrid/OrderListDatagrid.tsx:
```diff
@@ -37,32 +40,42 @@ export const OrderListDatagrid: React.FC<OrderListDatagridProps> = ({
+      columnCategories: [],
+      selectedColumns: settings?.columns ?? [],
+      onSave: handleColumnChange,
+      setDynamicColumnSettings,
+      columnPickerSettings,
```
I've removed it - although column picker settings are going to be dropped anyway in #3686 which is next issue I'm going to work on as soon as this is merged

Droniu commented on src/components/ColumnPicker/NewColumnPicker.tsx:
```diff
@@ -0,0 +1,122 @@
+      }}
+    >
+      <Popover.Trigger>
+        <Button
+          variant="tertiary"
+          icon={<TableEditIcon />}
+          __backgroundColor={
+            pickerOpen
+              ? vars.colors.background.interactiveNeutralSecondaryPressing
+              : undefined
+          }
+          __borderColor={
+            pickerOpen ? vars.colors.border.neutralSubdued : undefined
+          }
+        />
```
It's single component - I'd prefer to leave it as it is

andrzejewsky commented on src/components/ColumnPicker/NewColumnPickerAvailableNodes.tsx:
```diff
@@ -0,0 +1,84 @@
+  const areNodesLoading = currentCategory.availableNodes === undefined;
+  const areNodesEmpty = currentCategory.availableNodes?.length === 0;
+
+  const renderNodes = () => {
```
**suggestion**: this is basically a component, why do you want to be a function like this?

andrzejewsky replied:
Also you this this entire component when it's split into smaller pieces, can avoid props drilling - please notice that some you require some of the properties just for one use that even is happening during some of conditions (not always)

andrzejewsky commented on src/components/ColumnPicker/NewColumnPickerCategories.tsx:
```diff
@@ -0,0 +1,120 @@
+  onDynamicColumnSelect,
+  columnPickerSettings,
+}) => {
+  const [selectedCategory, setSelectedCategory] = useState<string | null>(null);
+
+  const currentCategory = useMemo(
+    () => columnCategories.find(category => category.name === selectedCategory),
+    [columnCategories, selectedCategory],
+  );
+
+  const [query, setQuery] = useState<string>("");
+
+  const changeHandler = (column: string) =>
+    columnPickerSettings.includes(column)
+      ? onDynamicColumnSelect(
+          columnPickerSettings.filter(currentCol => currentCol !== column),
+        )
+      : onDynamicColumnSelect([...columnPickerSettings, column]);
+
+  React.useEffect(() => {
+    // Preselect category when there is only one
+    if (columnCategories.length === 1) {
+      setSelectedCategory(columnCategories[0].name);
+    }
+  }, [columnCategories]);
+
+  React.useEffect(() => {
+    if (currentCategory) {
+      setQuery(currentCategory.initialSearch ?? "");
+    }
+  }, [currentCategory]);
+
```
suggestions: move this logic to the hooks, one responsibility each:

- selecting category -> one hook with your effect hook etc.
- query -> second hook with your effect hook etc.
- use them here

andrzejewsky commented on src/components/ColumnPicker/useColumns.ts:
```diff
@@ -0,0 +1,164 @@
+      columnCategories?.every(category => Array.isArray(category.selectedNodes))
+    ) {
+      updateDynamicColumns(
+        columnCategories
+          .flatMap(category => category.selectedNodes)
+          .sort(
+            (a, b) =>
+              columnPickerSettings.indexOf(a.id) -
+              columnPickerSettings.indexOf(b.id),
+          ),
```
**issue**: this is very hard to understand of what you want to do here, you can wrap it into function and that function name can indicate the meaning. Even you can split this function into smaller pieces, so the other developer can see a clear transition

andrzejewsky commented on src/components/ColumnPicker/useColumns.ts:
```diff
@@ -0,0 +1,164 @@
+
+  const initialColumnsState = React.useMemo(
+    () =>
+      [...staticColumns, ...(dynamicColumns ?? [])].filter(
+        column => selectedColumns.includes(column.id) || column.id === "empty",
+      ),
```
**issue**: similar here, dedicated functions + name them properly

andrzejewsky commented on src/components/ColumnPicker/useColumns.ts:
```diff
@@ -0,0 +1,164 @@
+
+  const onMove = React.useCallback(
+    (startIndex: number, endIndex: number): void => {
+      // When empty column prevent to rearrange it order
+      if (visibleColumns[0]?.id === "empty") {
+        if (startIndex === 0) {
+          return;
+        }
+
+        // Keep empty column always at beginning
+        if (endIndex === 0) {
+          return setVisibleColumns(old =>
+            addAtIndex(
+              old[startIndex],
+              removeAtIndex(old, startIndex),
+              endIndex + 1,
+            ),
+          );
+        }
+      }
+
+      setVisibleColumns(old =>
+        addAtIndex(old[startIndex], removeAtIndex(old, startIndex), endIndex),
+      );
```
**suggestion**: this is better, you left the comments, however it could be also a separated function

andrzejewsky commented on src/components/ColumnPicker/useColumns.ts:
```diff
@@ -0,0 +1,164 @@
+  );
+
+  const onResize = React.useCallback(
+    (column: GridColumn, newSize: number) => {
+      if (column.id === "empty") {
+        return;
+      }
+      return setVisibleColumns(prevColumns =>
+        prevColumns.map(prevColumn =>
+          prevColumn.id === column.id
+            ? { ...prevColumn, width: newSize }
+            : prevColumn,
+        ),
+      );
+    },
```
**issue**: I think you know.

andrzejewsky commented on src/components/ColumnPicker/useColumns.ts:
```diff
@@ -0,0 +1,164 @@
+    }
+    setDynamicColumnSettings(columns);
+    updateDynamicColumns(prevDynamicColumns =>
+      uniqBy(
+        [
+          ...(prevDynamicColumns ?? []),
+          ...columnCategories
+            .flatMap(category => category.availableNodes)
+            .filter(x => columns.includes(x.id)),
+        ],
+        "id",
+      )
+        .sort((a, b) => columns.indexOf(a.id) - columns.indexOf(b.id))
+        .filter(x => columns.includes(x.id)),
```
**issue**: guess what? :D

andrzejewsky commented on src/components/ColumnPicker/useColumns.ts:
```diff
@@ -0,0 +1,164 @@
+    columnPickerSettings,
+    recentlyAddedColumn,
+  };
+};
```
is this file yours? or you copied that? if so, please do very little refactoring if you can, I get that if it's not yours

poulch commented on src/components/Datagrid/ColumnPicker/ColumnPickerAvailableNodes.tsx:
```diff
@@ -0,0 +1,77 @@
+  query,
+  setQuery,
+  changeHandler,
+}) => {
```
```suggestion
}: ColumnPickerAvailableNodesProps) => {
```

poulch commented on src/components/Datagrid/ColumnPicker/ColumnPickerAvailableNodes.tsx:
```diff
@@ -0,0 +1,77 @@
+        </Box>
+      );
+    }
+    if (areNodesEmpty) {
```
suggestion: empty space before for better readability

poulch commented on src/components/Datagrid/ColumnPicker/ColumnPickerAvailableNodes.tsx:
```diff
@@ -0,0 +1,77 @@
+        </Text>
+      );
+    }
+    return currentCategory.availableNodes.map(node => (
```
suggestion: empty space before for better readability

poulch commented on src/components/Datagrid/ColumnPicker/ColumnPickerCategoryList.tsx:
```diff
@@ -0,0 +1,26 @@
+export const ColumnPickerCategoryList = ({
+  columnCategories,
+  setCurrentCategory,
+}) => (
```
```suggestion
}: ColumnPickerCategoryListProps) => (
```

poulch commented on src/components/Datagrid/ColumnPicker/ColumnPickerCategories.tsx:
```diff
@@ -0,0 +1,97 @@
+  onDynamicColumnSelect: (columns: string[]) => void;
+}
+
+export const ColumnPickerCategories: React.FC<ColumnPickerCategoriesProps> = ({
+  columnCategories,
+  onClose,
+  onDynamicColumnSelect,
+  columnPickerSettings,
+}) => {
```
```suggestion
export const ColumnPickerCategories = ({
  columnCategories,
  onClose,
  onDynamicColumnSelect,
  columnPickerSettings,
}: ColumnPickerCategoriesProps) => {
```

poulch commented on src/components/Datagrid/ColumnPicker/ColumnPicker.tsx:
```diff
@@ -0,0 +1,107 @@
+  onDynamicColumnSelect?: (columns: string[]) => void;
+}
+
+export const ColumnPicker: React.FC<ColumnPickerProps> = ({
+  staticColumns,
+  selectedColumns,
+  columnCategories,
+  dynamicColumns,
+  columnPickerSettings,
+  onDynamicColumnSelect,
+  onSave,
+}) => {
```
Keep the same convention everywhere
```suggestion
export const ColumnPicker = ({
  staticColumns,
  selectedColumns,
  columnCategories,
  dynamicColumns,
  columnPickerSettings,
  onDynamicColumnSelect,
  onSave,
}: ColumnPickerProps) => {
```

poulch commented on src/components/Datagrid/ColumnPicker/ColumnPickerDynamicColumns.tsx:
```diff
@@ -0,0 +1,51 @@
+  selectedColumns: string[];
+}
+
+export const ColumnPickerDynamicColumns: React.FC<
+  ColumnPickerDynamicColumnsProps
+> = ({ dynamicColumns, setExpanded, handleToggle, selectedColumns }) => (
```
```suggestion
export const ColumnPickerDynamicColumns= ({ 
dynamicColumns, 
setExpanded, 
handleToggle, 
selectedColumns 
}: ColumnPickerDynamicColumnsProps) => (
```

poulch commented on src/components/Datagrid/ColumnPicker/ColumnPickerSearch.tsx:
```diff
@@ -0,0 +1,43 @@
+  query: string;
+}
+
+export const ColumnPickerSearch: React.FC<ColumnPickerSearchProps> = ({
+  currentCategory,
+  setQuery,
+  query,
+}) => {
```
```suggestion
export const ColumnPickerSearch = ({
  currentCategory,
  setQuery,
  query,
}: ColumnPickerSearchProps) => {
```

poulch commented on src/components/Datagrid/ColumnPicker/ColumnPickerSearch.tsx:
```diff
@@ -0,0 +1,43 @@
+  query,
+}) => {
+  const intl = useIntl();
+  return (
```
suggestion: empty space before

poulch commented on src/components/Datagrid/ColumnPicker/ColumnPickerStaticColumns.tsx:
```diff
@@ -0,0 +1,28 @@
+  staticColumns,
+  handleToggle,
+  selectedColumns,
+}) =>
```
suggestion: add types here

poulch commented on src/components/Datagrid/ColumnPicker/useColumns.ts:
```diff
@@ -0,0 +1,166 @@
+    React.useState<AvailableColumn[]>(null);
+
+  React.useEffect(() => {
+    if (dynamicColumns === null && areCategoriesLoaded(columnCategories)) {
+      updateDynamicColumns(
+        sortColumns(
+          extractSelectedNodesFromCategories(columnCategories),
+          columnPickerSettings,
+        ),
+      );
+    }
```
suggestion: can you add comment what is going on here

poulch commented on src/components/Datagrid/ColumnPicker/useColumns.ts:
```diff
@@ -0,0 +1,166 @@
+    if (typeof setDynamicColumnSettings !== "function") {
+      return;
+    }
+    setDynamicColumnSettings(selected);
```
suggestion: empty space before

poulch reviewed (approved):
Good job 💪🏻  LGTM

wojteknowacki commented:
When too many dynamic columns are picked user is not able to see the whole list - the list goes down beyond the viewport and there is no scroll
![Zrzut ekranu 2023-06-13 o 09 18 31](https://github.com/saleor/saleor-dashboard/assets/124166231/26fd3a86-6b27-4703-8f64-fd7367cc0853)

wojteknowacki commented:
The user is not able to remove once selected dynamic columns from the custom section - in the long term, this will make the custom section a mess

wojteknowacki commented:
Inactive Toggle UI not visible when dark theme on 
<img width="639" alt="Zrzut ekranu 2023-06-13 o 11 11 38" src="https://github.com/saleor/saleor-dashboard/assets/124166231/440babc2-7072-4a38-85f0-ea7c5f325f96">

wojteknowacki commented:
Orders static column picker misbehaving - toggle states are not consistent 
https://github.com/saleor/saleor-dashboard/assets/124166231/1bbd087a-673e-4763-9e58-c3fd5609adf3

wojteknowacki commented:
Static column sate is not saved properly when moving between views 


https://github.com/saleor/saleor-dashboard/assets/124166231/28f702d1-69b5-4e9b-9ced-5f755adae117

wojteknowacki commented:
Column order is not kept when changing sorting

https://github.com/saleor/saleor-dashboard/assets/124166231/24c0eb62-e571-4720-9cb9-a0a83f33744a

wojteknowacki commented:
Ned elipsis or other solution in case of long attr name
![image](https://github.com/saleor/saleor-dashboard/assets/124166231/1e49ad8b-b33e-44d1-98ad-ab1613fc710a)
![image](https://github.com/saleor/saleor-dashboard/assets/124166231/2747f25a-eb7e-4cbc-a4ce-e7d02629ce79)

Droniu commented:
# ⭐ QA summary
 This is a quick review of the issues discovered during the QA process and how we plan to resolve them.
cc @wojteknowacki 
## 1. Picker overflowing for large lists
Fixed 76b4112233f57b390f53c79c299edafe39dad936
Min and max height has been added to account for various screen resolutions.
## 2. Removing dynamic columns from right side of picker
Issue #3686.
This is going to be worked on right after we merge this PR.
## 3. Toggles have bad contrast on dark mode
We have the designs for toggles with white dots already, but we need to create a new token. This can be done separately and will be tracked in https://github.com/saleor/macaw-ui/issues/471
## 4. Hover effect on full width of the labels in candidate columns
Fixed ad452eac39b876c0d8618e775b7be1a4209dd94e
Checkboxes now span full width - both the hover effect is visible and they are clickable.
## 5. Bad contrast of disabled toggles
This is under discussion with the design team.
## 6. State of static columns is lost after changing views
WIP
## 7. Order of the column is not persisted when sorting by column.
Although this is a slight regression, this happens due to architectural changes in this PR. This is going to be restored *probably* in #3686 - if not, this is going to be *definitely* restored in #3682 where we implement full persistence - order of the columns is going to be saved in LS, so even after refreshing and navigating to other places it's gonna stay constant.

## 8. Column picker is ultrawide when used with long attribute names.
I've added ellipsis. WIP 

## 9. Closing the picker by clicking on the datagrid navigates to product details.
WIP

## 10. `data-test-id` props
WIP

poulch commented on src/products/queries.ts:
```diff
@@ -250,14 +249,54 @@ export const productMediaQuery = gql`
-  query GridAttributes($ids: [ID!]!) {
-    grid: attributes(first: 25, filter: { ids: $ids }) {
+  query GridAttributes($ids: [ID!]!, $hasAttributes: Boolean!) {
+    left: attributes(first: 10) {
```
suggestion: using `left | right` doesn't mean what this data is, also it's not good practice to suggest that this part of query will be use in left side of ui.

Droniu commented on src/components/Datagrid/ColumnPicker/utils.tsx:
```diff
@@ -0,0 +1,104 @@
+  selectedColumns: string[];
+}) =>
+  [...staticColumns, ...(dynamicColumns ?? [])].filter(
+    column => selectedColumns.includes(column.id) || column.id === "empty",
```
Actually we cannot reuse `filterSelectedColumns` here because predicate also takes empty column into the account.

Droniu commented on src/hooks/useListSettings.ts:
```diff
@@ -12,6 +12,13 @@ export interface UseListSettings<TColumns extends string = string> {
   ) => void;
 }
+
+const mergeCustomizer = (objValue: unknown, srcValue: unknown) => {
+  if (Array.isArray(objValue) && Array.isArray(srcValue)) {
+    return srcValue;
+  }
+};
+
```
We have a unit test for this hook already written - it passes, problems could only occur when deeply-nested array values are added to `listConfig` in the future, but I believe this will be detected easily. I've also added a comment.
